cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -0,0 +1,1100 @@
+
+    auto outcome = client_->AbortMultipartUpload(req);
+    if (!outcome.IsSuccess()) {
+      return ErrorToStatus(outcome.GetError());
+    }
+    current_part_.reset();
+    closed_ = true;
+    return Status::OK();
+  }
+
+  // OutputStream interface
+
+  Status Close() override {
+    if (closed_) {
+      return Status::OK();
+    }
+
+    if (current_part_) {
+      // Upload last part
+      RETURN_NOT_OK(CommitCurrentPart());
+    }
+
+    // S3 mandates at least one part, upload an empty one if necessary
+    if (completed_upload_.GetParts().empty()) {
+      RETURN_NOT_OK(UploadPart("", 0));
+    }
+    DCHECK(completed_upload_.PartsHasBeenSet());
+
+    S3Model::CompleteMultipartUploadRequest req;
+    req.SetBucket(ToAwsString(path_.bucket));
+    req.SetKey(ToAwsString(path_.key));
+    req.SetUploadId(upload_id_);
+    req.SetMultipartUpload(completed_upload_);
+
+    auto outcome = client_->CompleteMultipartUpload(req);
+    if (!outcome.IsSuccess()) {
+      return ErrorToStatus(outcome.GetError());
+    }
+
+    closed_ = true;
+    return Status::OK();
+  }
+
+  bool closed() const override { return closed_; }
+
+  Status Tell(int64_t* position) const override {
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+    *position = pos_;
+    return Status::OK();
+  }
+
+  Status Write(const void* data, int64_t nbytes) override {
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+
+    // With up to 10000 parts in an upload (S3 limit), a stream writing chunks
+    // of exactly 5MB would be limited to 50GB total.  To avoid that, we bump
+    // the upload threshold every 100 parts.  So the pattern is:
+    // - part 1 to 99: 5MB threshold
+    // - part 100 to 199: 10MB threshold
+    // - part 200 to 299: 15MB threshold
+    // ...
+    // - part 9900 to 9999: 500MB threshold
+    // So the total size limit is 2475000MB or ~2.4TB, while keeping manageable
+    // chunk sizes and avoiding too much buffering in the common case of a small-ish
+    // stream.  If the limit's not enough, we can revisit.
+    if (part_number_ % 100 == 0) {
+      part_upload_threshold_ += kMinimumPartUpload;
+    }
+
+    if (!current_part_ && nbytes >= part_upload_threshold_) {
+      // No current part and data large enough, upload it directly without copying
+      RETURN_NOT_OK(UploadPart(data, nbytes));
+      pos_ += nbytes;
+      return Status::OK();
+    }
+    // Can't upload data on its own, need to buffer it
+    if (!current_part_) {
+      RETURN_NOT_OK(io::BufferOutputStream::Create(
+          part_upload_threshold_, default_memory_pool(), &current_part_));
+      current_part_size_ = 0;
+    }
+    RETURN_NOT_OK(current_part_->Write(data, nbytes));
+    pos_ += nbytes;
+    current_part_size_ += nbytes;
+
+    if (current_part_size_ >= part_upload_threshold_) {
+      // Current part large enough, upload it
+      RETURN_NOT_OK(CommitCurrentPart());
```

Comment:
Good idea.

Reply:
This may be a more general need for other filesystem implementations, so perhaps there would be merit in putting such functionality inside `BufferedOutputStream`.

Reply:
@fsaintjacques Not really. Neither `BufferOutputStream` nor `BufferBuilder` allows to re-use the same buffer after finishing. Also this wouldn't be compatible with background writes.